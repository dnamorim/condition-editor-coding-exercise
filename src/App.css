.container {
  display: block;
  margin: 5vw;
}

.filters {
  display: flex;
  flex-flow: row wrap;
  gap: 3em;
}

.filters .filters-item {
  min-width: 14em;
  min-height: 2.71428571em;
  flex-grow: 0;
}

.filters .dynamic-width {
  flex-grow: 2;
}

.clearButton {
  float: right;
}

@media all and (max-width: 1010px) {
  .filterList {
    display: flex;
    flex-direction: column;
    row-gap: 1em;
  }

  .clearButton {
    width: 100%;
  }
}
