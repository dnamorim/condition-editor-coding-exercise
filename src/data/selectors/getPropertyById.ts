const getPropertyById = (propertyId: string) =>
  window.datastore
    .getProperties()
    .find(
      (property) => propertyId !== '' && property.id === Number(propertyId),
    );

export default getPropertyById;
