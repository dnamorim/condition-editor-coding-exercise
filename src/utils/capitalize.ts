const capitalize = (text: string | undefined) => {
  if (!text) {
    return text;
  }

  return text[0].toUpperCase() + text.slice(1).toLowerCase();
};

export default capitalize;
