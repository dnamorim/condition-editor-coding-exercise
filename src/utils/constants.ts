/* Property Types */
export const PROPERTY_TYPES = {
  STRING: 'string',
  NUMBER: 'number',
  ENUM: 'enumerated',
};

/* Operator Types */
export const OPERATOR_TYPES = {
  EQUALS: 'equals',
  GREATER_THAN: 'greater_than',
  LESS_THAN: 'less_than',
  ANY: 'any',
  NONE: 'none',
  IN: 'in',
  CONTAINS: 'contains',
};
