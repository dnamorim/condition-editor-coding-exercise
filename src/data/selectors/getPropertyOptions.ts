import { DropdownItemProps } from 'semantic-ui-react';
import capitalize from 'utils/capitalize';

const getPropertyOptions = (): DropdownItemProps[] =>
  window.datastore.getProperties().map((property) => ({
    key: property.id,
    text: capitalize(property.name),
    value: String(property.id),
  }));

export default getPropertyOptions;
