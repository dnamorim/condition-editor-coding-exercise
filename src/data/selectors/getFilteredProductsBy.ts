import { Property, PropertyValue } from 'data/datastore.types';
import { OPERATOR_TYPES, PROPERTY_TYPES } from 'utils/constants';
import { ConditionFilterType, ProductListItem } from './types';

const getPropertyValue = (
  propertyValues: PropertyValue[],
  selectedProperty: number,
) => propertyValues.find(({ property_id }) => property_id === selectedProperty);

const getConditionFilter = (
  selectedProperty: Property,
  selectedOperator: string,
) => {
  let conditionFilter: ConditionFilterType = () => true;

  switch (selectedOperator) {
    case OPERATOR_TYPES.GREATER_THAN:
      conditionFilter = (value, filter) => Number(value) > Number(filter);
      break;
    case OPERATOR_TYPES.LESS_THAN:
      conditionFilter = (value, filter) => Number(value) < Number(filter);
      break;
    case OPERATOR_TYPES.EQUALS:
      conditionFilter = (value, filter) =>
        selectedProperty.type === PROPERTY_TYPES.NUMBER
          ? Number(value) === Number(filter)
          : value === filter;
      break;
    case OPERATOR_TYPES.CONTAINS:
      conditionFilter = (value, filter) => value.includes(filter);
      break;
    case OPERATOR_TYPES.IN:
      conditionFilter = (value, filter) =>
        selectedProperty.type === PROPERTY_TYPES.ENUM
          ? filter?.includes(value) || false
          : filter?.split(/;|,/).includes(`${value}`) ?? false;
      break;
    case OPERATOR_TYPES.ANY:
      conditionFilter = (value, _) => !!value;
      break;
    case OPERATOR_TYPES.NONE:
      conditionFilter = (value, _) => value === null || value === undefined;
      break;
    default:
      conditionFilter = () => true;
      break;
  }

  return conditionFilter;
};

const getFilteredProductsBy = (
  selectedProperty: Property | undefined,
  selectedOperator: string,
  valueFilter?: string | string[] | undefined,
): ProductListItem[] => {
  const products = window.datastore.getProducts();
  let productList = products;

  if (selectedProperty !== undefined && selectedOperator !== '') {
    const formattedFilter =
      typeof valueFilter === PROPERTY_TYPES.STRING
        ? (valueFilter as string).toLowerCase()
        : valueFilter;

    const conditionFilter = getConditionFilter(
      selectedProperty,
      selectedOperator,
    );

    productList = productList.filter((product) => {
      let propertyValue = getPropertyValue(
        product.property_values,
        selectedProperty.id,
      )?.value;

      propertyValue =
        typeof propertyValue === PROPERTY_TYPES.STRING
          ? propertyValue?.toLowerCase()
          : propertyValue;

      return conditionFilter(propertyValue, formattedFilter);
    });
  }

  return productList.map((product) => {
    const properties = product.property_values.reduce(
      (dictionary, propertyValue) => {
        dictionary[propertyValue.property_id] = propertyValue.value;
        return dictionary;
      },
      {} as Record<number, string>,
    );

    return {
      id: product.id,
      properties,
    };
  });
};

export default getFilteredProductsBy;
