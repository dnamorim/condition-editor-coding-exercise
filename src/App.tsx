import { useMemo, useState } from 'react';
import { Button, DropdownItemProps } from 'semantic-ui-react';

import ProductList from 'components/ProductList';
import FilterDropdown from 'components/FilterDropdown';
import ValueFilter from 'components/ValueFilter';

import useDebounce from 'utils/useDebounce';

import getFilteredProductsBy from 'data/selectors/getFilteredProductsBy';
import getPropertyOptions from 'data/selectors/getPropertyOptions';
import getOperationOptions from 'data/selectors/getOperationsOptions';
import getPropertyById from 'data/selectors/getPropertyById';
import { ProductListItem } from 'data/selectors/types';

import 'data/datastore';
import './App.css';

function App() {
  const [selectedPropertyId, setPropertyId] = useState('');
  const [selectedOperatorId, setOperatorId] = useState('');
  const [valueFilter, setValueFilter] = useState('');
  const debouncedValueFilter = useDebounce(valueFilter, 400);

  const propertyOptions = getPropertyOptions();
  const selectedProperty = useMemo(
    () => getPropertyById(selectedPropertyId),
    [selectedPropertyId],
  );
  const operationOptions = useMemo<DropdownItemProps[]>(
    () => getOperationOptions(selectedPropertyId),
    [selectedPropertyId],
  );

  const products = useMemo<ProductListItem[]>(
    () =>
      getFilteredProductsBy(
        selectedProperty,
        selectedOperatorId,
        debouncedValueFilter,
      ),
    [debouncedValueFilter, selectedProperty, selectedOperatorId],
  );

  const handleOnChangeProperty = (_: any, { value }: any) => {
    setPropertyId(value);
    setValueFilter('');
  };

  const handleOnChangeOperator = (_: any, { value }: any) => {
    setOperatorId(value);
  };

  const handleOnChangeValueFilter = (_: any, { value }: any) => {
    setValueFilter(value);
  };

  const handleOnClick = () => {
    setPropertyId('');
    setOperatorId('');
    setValueFilter('');
  };

  return (
    <div className="container">
      <div className="filters">
        <div className="filters-item">
          <FilterDropdown
            data-testid="propertyFilter"
            placeholder="Property"
            options={propertyOptions}
            onChange={handleOnChangeProperty}
            value={selectedPropertyId}
          />
        </div>
        <div className="filters-item">
          <FilterDropdown
            data-testid="operatorFilter"
            placeholder="Operator"
            disabled={!selectedProperty}
            options={operationOptions}
            onChange={handleOnChangeOperator}
            value={selectedOperatorId}
          />
        </div>
        <div className="filters-item dynamic-width">
          <ValueFilter
            data-testid="valueFilter"
            placeholder="Value(s)"
            property={selectedProperty}
            operator={selectedOperatorId}
            onChange={handleOnChangeValueFilter}
            value={valueFilter}
          />
        </div>
        <div className="item dynamic-width">
          <Button
            data-testid="clearButton"
            className="clearButton"
            onClick={handleOnClick}
          >
            Clear
          </Button>
        </div>
      </div>

      <ProductList properties={propertyOptions} products={products} />
    </div>
  );
}

export default App;
