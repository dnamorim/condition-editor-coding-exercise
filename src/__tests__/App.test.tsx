import { fireEvent, render, screen, within } from '@testing-library/react';
import getFilteredProductsBy from 'data/selectors/getFilteredProductsBy';
import getPropertyOptions from 'data/selectors/getPropertyOptions';
import getOperationOptions from 'data/selectors/getOperationsOptions';
import getPropertyById from 'data/selectors/getPropertyById';
import { PROPERTY_TYPES } from 'utils/constants';

import App from '../App';

jest.mock('data/selectors/getFilteredProductsBy');
jest.mock('data/selectors/getOperationsOptions');
jest.mock('data/selectors/getPropertyOptions');
jest.mock('data/selectors/getPropertyById');
// @ts-expect-error error
jest.mock('utils/useDebounce.ts', () => (value) => value);

const mockPropertyOptions = [
  {
    key: 0,
    text: 'Name',
    value: '0',
  },
  {
    key: 1,
    text: 'Weight',
    value: '1',
  },
];

const mockOperationOptions = [
  {
    key: 'equals',
    text: 'Equals',
    value: 'equals',
  },
  {
    key: 'contains',
    text: 'Contains',
    value: 'contains',
  },
];

const mockProductList = [
  { id: 111, properties: { 0: 'Keyboard', 1: 5 } },
  { id: 222, properties: { 0: 'Cup', 1: 3 } },
];

const mockProperty = {
  id: mockPropertyOptions[0].key,
  name: mockOperationOptions[0].text,
  type: PROPERTY_TYPES.STRING,
};

describe('App', () => {
  beforeEach(jest.clearAllMocks);

  it('should render correcty when no products are available', () => {
    // @ts-expect-error error
    getFilteredProductsBy.mockReturnValue([]);
    // @ts-expect-error error
    getPropertyOptions.mockReturnValue(mockPropertyOptions);
    // @ts-expect-error error
    getOperationOptions.mockReturnValue([]);

    const { container } = render(<App />);

    const productRows = screen.getAllByRole('rowgroup')[1].childNodes;

    expect(productRows).toHaveLength(0);
    expect(screen.getByTestId('propertyFilter')).toBeInTheDocument();
    expect(screen.getByTestId('operatorFilter')).toBeInTheDocument();
    expect(screen.getByTestId('clearButton')).toBeInTheDocument();
    expect(screen.queryByTestId('valueFilter')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render a list of products correctly', () => {
    // @ts-expect-error error
    getFilteredProductsBy.mockReturnValue(mockProductList);
    // @ts-expect-error error
    getPropertyOptions.mockReturnValue(mockPropertyOptions);
    // @ts-expect-error error
    getOperationOptions.mockReturnValue([]);

    const { container } = render(<App />);

    const productRows = screen.getAllByRole('rowgroup')[1].childNodes;

    expect(productRows).toHaveLength(mockProductList.length);
    expect(
      screen.getByText(mockProductList[0].properties[0]),
    ).toBeInTheDocument();
    expect(
      screen.getByText(mockProductList[1].properties[0]),
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should allow operator selection when the property is set', () => {
    // @ts-expect-error error
    getFilteredProductsBy.mockReturnValue([]);
    // @ts-expect-error error
    getPropertyOptions.mockReturnValue(mockPropertyOptions);
    // @ts-expect-error error
    getOperationOptions.mockReturnValue(mockOperationOptions);
    // @ts-expect-error error
    getPropertyById.mockReturnValueOnce().mockReturnValueOnce(mockProperty);

    render(<App />);

    expect(screen.getByTestId('operatorFilter')).toHaveAttribute(
      'aria-disabled',
      'true',
    );

    const propertyElement = screen.getByTestId('propertyFilter');
    fireEvent.click(screen.getByTestId('propertyFilter'));
    const propertyOptions = within(propertyElement).getAllByRole('option');
    fireEvent.click(propertyOptions[0]);

    expect(screen.getByTestId('operatorFilter')).toHaveAttribute(
      'aria-disabled',
      'false',
    );
  });

  it('should display the value filter when the property and operator are set', () => {
    // @ts-expect-error error
    getFilteredProductsBy.mockReturnValue([]);
    // @ts-expect-error error
    getPropertyOptions.mockReturnValue(mockPropertyOptions);
    // @ts-expect-error error
    getOperationOptions.mockReturnValue(mockOperationOptions);
    // @ts-expect-error error
    getPropertyById.mockReturnValue(mockProperty);

    const { container } = render(<App />);

    expect(screen.queryByTestId('valueFilter')).not.toBeInTheDocument();

    const propertyElement = screen.getByTestId('propertyFilter');
    fireEvent.click(screen.getByTestId('propertyFilter'));
    const propertyOptions = within(propertyElement).getAllByRole('option');
    fireEvent.click(propertyOptions[0]);

    const operatorElement = screen.getByTestId('operatorFilter');
    fireEvent.click(operatorElement);
    const operationOptions = within(operatorElement).getAllByRole('option');
    fireEvent.click(operationOptions[0]);

    expect(screen.getByTestId('valueFilter')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should fetch products when all the filters are selected', () => {
    // @ts-expect-error error
    getFilteredProductsBy.mockReturnValue([]);
    // @ts-expect-error error
    getPropertyOptions.mockReturnValue(mockPropertyOptions);
    // @ts-expect-error error
    getOperationOptions.mockReturnValue(mockOperationOptions);
    // @ts-expect-error error
    getPropertyById.mockReturnValue(mockProperty);

    render(<App />);

    const propertyElement = screen.getByTestId('propertyFilter');
    fireEvent.click(screen.getByTestId('propertyFilter'));
    const propertyOptions = within(propertyElement).getAllByRole('option');
    fireEvent.click(propertyOptions[0]);

    const operatorElement = screen.getByTestId('operatorFilter');
    fireEvent.click(operatorElement);
    const operationOptions = within(operatorElement).getAllByRole('option');
    fireEvent.click(operationOptions[0]);

    const mockValueFilter = 'foo';
    const valueElement = within(screen.getByTestId('valueFilter')).getByRole(
      'textbox',
    );
    fireEvent.change(valueElement, { target: { value: mockValueFilter } });

    expect(getFilteredProductsBy).toHaveBeenLastCalledWith(
      mockProperty,
      mockOperationOptions[0].value,
      mockValueFilter,
    );
  });

  it('should reset all filters and show all products', () => {
    const filteredList = [mockProductList[1]];
    let productRows;

    // @ts-expect-error error
    getFilteredProductsBy.mockReturnValue(filteredList);
    // @ts-expect-error error
    getPropertyOptions.mockReturnValue(mockPropertyOptions);
    // @ts-expect-error error
    getOperationOptions.mockReturnValue(mockOperationOptions);
    // @ts-expect-error error
    getPropertyById.mockReturnValue(mockProperty);

    render(<App />);

    const propertyElement = screen.getByTestId('propertyFilter');
    fireEvent.click(screen.getByTestId('propertyFilter'));
    const propertyOptions = within(propertyElement).getAllByRole('option');
    fireEvent.click(propertyOptions[0]);

    const operatorElement = screen.getByTestId('operatorFilter');
    fireEvent.click(operatorElement);
    const operationOptions = within(operatorElement).getAllByRole('option');
    fireEvent.click(operationOptions[0]);

    const mockValueFilter = 'foo';
    const valueElement = within(screen.getByTestId('valueFilter')).getByRole(
      'textbox',
    );
    fireEvent.change(valueElement, { target: { value: mockValueFilter } });

    productRows = screen.getAllByRole('rowgroup')[1].childNodes;

    expect(productRows).toHaveLength(filteredList.length);

    // @ts-expect-error error
    getFilteredProductsBy.mockReturnValueOnce(mockProductList);

    const clearButton = screen.getByTestId('clearButton');
    fireEvent.click(clearButton);

    productRows = screen.getAllByRole('rowgroup')[1].childNodes;

    expect(productRows).toHaveLength(mockProductList.length);
    expect(
      within(screen.getByTestId('propertyFilter')).getByText('Property'),
    ).toBeInTheDocument();
    expect(
      within(screen.getByTestId('operatorFilter')).getByText('Operator'),
    ).toBeInTheDocument();
    expect(screen.queryByTestId('valueFilter')).not.toBeInTheDocument();
  });
});
