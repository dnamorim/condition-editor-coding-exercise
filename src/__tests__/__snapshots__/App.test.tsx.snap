// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should display the value filter when the property and operator are set 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="filters"
    >
      <div
        class="filters-item"
      >
        <div
          aria-expanded="false"
          class="ui fluid selection dropdown"
          data-testid="propertyFilter"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            Name
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="true"
              aria-selected="true"
              class="active selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Name
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Weight
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="filters-item"
      >
        <div
          aria-disabled="false"
          aria-expanded="true"
          class="ui active visible fluid selection dropdown"
          data-testid="operatorFilter"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            Equals
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="visible menu transition"
          >
            <div
              aria-checked="true"
              aria-selected="true"
              class="active selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Equals
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Contains
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="filters-item dynamic-width"
      >
        <div
          class="ui fluid input"
          data-testid="valueFilter"
        >
          <input
            placeholder="Value(s)"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="item dynamic-width"
      >
        <button
          class="ui button clearButton"
          data-testid="clearButton"
        >
          Clear
        </button>
      </div>
    </div>
    <table
      class="ui celled striped table"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
          >
            Name
          </th>
          <th
            class=""
          >
            Weight
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      />
    </table>
  </div>
</div>
`;

exports[`App should render a list of products correctly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="filters"
    >
      <div
        class="filters-item"
      >
        <div
          aria-expanded="false"
          class="ui fluid selection dropdown"
          data-testid="propertyFilter"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider default text"
            role="alert"
          >
            Property
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="true"
              class="selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Name
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Weight
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="filters-item"
      >
        <div
          aria-disabled="true"
          aria-expanded="false"
          class="ui disabled fluid selection dropdown"
          data-testid="operatorFilter"
          role="listbox"
          tabindex="-1"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider default text"
            role="alert"
          >
            Operator
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          />
        </div>
      </div>
      <div
        class="filters-item dynamic-width"
      />
      <div
        class="item dynamic-width"
      >
        <button
          class="ui button clearButton"
          data-testid="clearButton"
        >
          Clear
        </button>
      </div>
    </div>
    <table
      class="ui celled striped table"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
          >
            Name
          </th>
          <th
            class=""
          >
            Weight
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
        >
          <td
            class=""
          >
            Keyboard
          </td>
          <td
            class=""
          >
            5
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class=""
          >
            Cup
          </td>
          <td
            class=""
          >
            3
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`App should render correcty when no products are available 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="filters"
    >
      <div
        class="filters-item"
      >
        <div
          aria-expanded="false"
          class="ui fluid selection dropdown"
          data-testid="propertyFilter"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider default text"
            role="alert"
          >
            Property
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="true"
              class="selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Name
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Weight
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="filters-item"
      >
        <div
          aria-disabled="true"
          aria-expanded="false"
          class="ui disabled fluid selection dropdown"
          data-testid="operatorFilter"
          role="listbox"
          tabindex="-1"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider default text"
            role="alert"
          >
            Operator
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          />
        </div>
      </div>
      <div
        class="filters-item dynamic-width"
      />
      <div
        class="item dynamic-width"
      >
        <button
          class="ui button clearButton"
          data-testid="clearButton"
        >
          Clear
        </button>
      </div>
    </div>
    <table
      class="ui celled striped table"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
          >
            Name
          </th>
          <th
            class=""
          >
            Weight
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      />
    </table>
  </div>
</div>
`;
